### PR TITLE
relu_mlp updates

### PR DESCRIPTION
**Pull request by runjerry**
* fix a bug in direct Jacobian computation when input batch_size is 1

* add direct vector-Jacobian product computation

### CODE REVIEW

File: alf/networks/relu_mlp.py [hunk 1]
```diff
@@ -179,3 +178,45 @@ def _compute_jac_diag(self):
...
+    def _compute_vjp(self, vec):
+        """Compute vector-Jacobian product. """
+
+        if vec.ndim == 1:
+            vec = vec.unsqueeze(0)
+
+        J = torch.matmul(vec, self._fc_layers[-1].weight)
+        for fc in reversed(self._fc_layers[0:-1]):
+            mask = (fc.hidden_neurons > 0).float()
+            J = torch.einsum('ba, ba, aj->bj', J, mask, fc.weight)
```
Comment (emailweixu) [inline alf/networks/relu_mlp.py]: I don't have very high confidence on the performance of einsum. matmul is generally highly optimized. And hence ``matmul(J * mask, fc.weight)`` might be faster.

File: alf/networks/relu_mlp.py [hunk 2]
```diff
@@ -112,9 +112,10 @@ def forward(self,
             requires_jac (bool): whether outputs input-output Jacobian.
             requires_jac_diag (bool): whetheer outputs diagonals of Jacobian.
         """
-        inputs = inputs.squeeze()
-        assert inputs.shape[-1] == self._input_size, \
-            ("inputs should has shape {}!".format(self._input_size))
+        if inputs.ndim == 1:
```
Comment (emailweixu) [inline alf/networks/relu_mlp.py]: The shape of inputs and output are not consistent when ndim==1. Should squeeze the output if ndim==1

File: alf/networks/relu_mlp.py [hunk 3]
```diff
@@ -179,3 +178,45 @@ def _compute_jac_diag(self):
...
+
+        if vec.ndim == 1:
+            vec = vec.unsqueeze(0)
+
+        J = torch.matmul(vec, self._fc_layers[-1].weight)
+        for fc in reversed(self._fc_layers[0:-1]):
+            mask = (fc.hidden_neurons > 0).float()
+            J = torch.einsum('ba, ba, aj->bj', J, mask, fc.weight)
+
+        return J  # [B, n_in]
```
Comment (emailweixu) [inline alf/networks/relu_mlp.py]: input and output shape inconsistency.

File: alf/networks/relu_mlp.py (hunk 2 above)
Comment (runjerry) [reply alf/networks/relu_mlp.py]: updated.

File: alf/networks/relu_mlp.py (hunk 3 above)
Comment (runjerry) [reply alf/networks/relu_mlp.py]: updated.

File: alf/networks/relu_mlp.py (hunk 1 above)
Comment (runjerry) [reply alf/networks/relu_mlp.py]: updated.

File: alf/networks/relu_mlp.py [hunk 4]
```diff
@@ -134,8 +137,11 @@ def compute_jac(self, inputs):
             ("inputs should has shape {}!".format(self._input_size))
 
         self.forward(inputs)
+        J = self._compute_jac()
+        if inputs.ndim == 1:
+            J = J.squeeze()
```
Comment (emailweixu) [inline alf/networks/relu_mlp.py]: To be safe, use J.squeeze(0) in case there are some other singleton dimensions,

Comment (runjerry) [reply alf/networks/relu_mlp.py]: Updated here and other places.

**Review by emailweixu — APPROVED** [review]

**Pull request merged**